docs/client-design.md

```diff
@@ -0,0 +1,287 @@
+class Connection {
+public:
+    Connection(IOEventProvider &net_provider, 
+	           const ConnectionOptions &connection_options);
+	
+    /** 
+	 * Return the connection's status.
+	 */
+	enum class ConnectionStatus get_status() const;
+	
+	/** 
+	 * If the connection is an erroneous state (see `status`), return the 
+	 * connection error that caused it. The same error is also passed to 
+	 * response callbacks, and the same error is returned by 
+	 * `Request::get_error`. 
+	 */
+	std::optional<ConnectionError> &get_error() & const;
+	
+	/* An abstract request's interface. A request object is always returned. */
+	Request some_request(/* options */);
+};
+```
+
+#### Connect Options
+
+```c++
+/* Extend `ConnectionOptions` with an option for the reconnection feature. */
+struct ConnectOptions {
+	/* All existing options. */
+	
+	/** 
+	 * In the event of a broken connection, the interval in which the stream 
+	 * tries to re-establish the connection. 
+	 */
+	static constexpr size_t DEFAULT_RECONNECTION_INTERVAL = 2;
+	size_t reconnection_interval = DEFAULT_RECONNECTION_INTERVAL;
+};
+```
+
+### Requests
+
+```c++
+/** Encapsulates management of a request issued through a connection. */
+class Request {
+public:
+	/** 
+	 * Callback called when response is ready. Since the callback has a fixed 
+	 * signature, we need to allow for capturing additional context using 
+	 * lambdas. Hence, we use `std::function` for type erasure.
+	 * 
+	 * See `Connection::get_error` for details about the `error` parameter.
+	 */ 
+	using request_cb_f = 
+		std::function<void(const std::optional<ConnectionError> &error, 
+		                   Response &&response, Connection &connection)>;
+	
+	/** Set a callback called when the response is ready. */
+	int set_callback(request_cb_f request_cb);
+	
+	/** Remove the callback associated with the request, if any. */
+	void reset_callback();
+	
+	/** Return the request's status. */
+	enum class RequestStatus get_status();
+	
+	/**
+	 * Return a connection error, if any. See `Connection::get_error` for 
+	 * details.
+	 */
+	std::optional<ConnectionError> &get_error() & const;
+	
+	/**
+	 * Cancel the request, ending the lifetime of the corresponding response.
+	 */
+	void cancel();
+	
+	/** 
+	 * Return the response, if any. A response is available, iff:
+	 * 1. The response has been received by the connection.
+	 * 2. The response has not been dispatched to a callback.
+	 * 3. The response has not already been retrieved previously.
+	 */
+	std::optional<Response> get_response();
+};
+```
+
+#### Request Fan-Out
+
+```c++
+/**
+ * A fan-out encapsulates the collection of responses for a collection of 
+ * requests.
+ */
+class FanOut {
+	template<class InputIt>
+	FanOut(InputIt first, InputIt last)
+	FanOut(std::initializer_list<Request> requests);
+	
+	/**
+	 * Return a list of responses, and cancel the requests for which a response 
+	 * is not available. 
+	 */
+	tnt::List<Response> collect();
```

Comment:
How should we report errors here?

Reply:
Maybe, we should introduce `Result`, which is `variant<Error, Response>`, with methods `is_ok`, `get_error`, `get_response`.

Reply:
Actually, since we want the requests to be linked in. an intrusive list, I guess we should return a list of requests instead.

Reply:
What's the problem to have a list of Results?

Reply:
If we had an intrusive list of Results, we would skip one indirection level.

Reply:
> If we had an intrusive list of Results, we would skip one indirection level.

Not clear to me what you mean. Doesn't `std::variant` also introduce indirection?